[DennieDan] iP

DennieDan opened the pull request:
# Tecna
> Bring **magic** to all your tasks 🪄 

* You feel tired seeing the following tasks?
- [ ] Finish Milestone 3
- [ ] Go training at 6 AM
- [ ] Code review
* You don't know how to start a Java project, even a single ``main()`` method as below?
```Java
public class Tecna {
    public static void main(String[] args) {
        System.out.println("Hi, I'm Tecna!");
    }
}
```
### It means you _must_ try ***Tecna***

*To know more about Tecna, you can refer to the following pages:*
1. Tecna's [introduction](https://www.winxclub.com/en/article/introducing-myself-i%E2%80%99m-tecna)
2. Tecna's [best moments](https://www.youtube.com/watch?v=7TdIdWjJPuU)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DennieDan committed:
refactor: rename file from Duke to Tecna

DennieDan committed:
build: echo method

DennieDan committed:
build: add and list functions

DennieDan committed:
build: Task class in A-classes extension

DennieDan committed:
fix: add and list

DennieDan committed:
build: done and undone functions in main

DennieDan committed:
refactor: change variable name "todo" -> "taskList"
prevent confusion

DennieDan committed:
build: subclasses of Task

DennieDan committed:
build: input string parser

DennieDan committed:
refactor: Tecna.java reconstruct addItem() method

DennieDan committed:
config: input.txt and runtest.bat for textui tests

DennieDan committed:
build: InvalidRequestException added

DennieDan committed:
build: handle todo empty description

DennieDan committed:
build: delete function with O(n)

DennieDan committed:
add: delete test case

DennieDan committed:
update: java array to ArrayList

DennieDan committed:
Update Deadline to store LocalDate

DennieDan committed:
Update Deadline and Event to store LocalDateTime object

DennieDan committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-Level-7

DennieDan committed:
Update Tecna to read from a json file

DennieDan committed:
Update Tecna to save changes to a file

DennieDan committed:
Add credit to TaskWriter class

DennieDan committed:
Merge branch 'branch-Level-7'

DennieDan committed:
Merge branch 'master' into branch-Level-8

# Conflicts:
#	src/main/java/Tecna.java

DennieDan committed:
Fix bugs for parsing Date and Time

DennieDan committed:
Merge branch 'branch-Level-8'

DennieDan committed:
Add Storage class

DennieDan committed:
Add TaskList class

DennieDan committed:
Add TaskList class (2)

DennieDan committed:
Add CommandScanner class

DennieDan committed:
Fix CommandScanner incompatibility

DennieDan committed:
Add Ui class

DennieDan committed:
Merge branch 'A-MoreOOP'

DennieDan committed:
Add all to Tecna package
Packages extension minimal

DennieDan committed:
Merge branch 'A-Packages'

DennieDan committed:
Modify application plugin in build.gradle

DennieDan committed:
Merge branch 'A-Gradle'

DennieDan committed:
Add CommandScanner getRequest unit test

DennieDan committed:
Build Storage TaskParseException test

DennieDan committed:
Build Storage TaskParseException test

DennieDan committed:
Handle parsing exception in Storage class

DennieDan committed:
Delete TaskParseException

DennieDan committed:
Handle exceptions parsing tasks from json

DennieDan committed:
Test Storage class

DennieDan committed:
Handle number format in user command

DennieDan committed:
Handle deadline and event errors

DennieDan committed:
Test CommandScanner getRequest class

DennieDan committed:
Merge branch 'A-JUnit'

DennieDan committed:
Create a Jar file

DennieDan committed:
Add JavaDoc for Tecna class

DennieDan committed:
Add JavaDoc for CommandScanner

DennieDan committed:
Add JavaDoc to 2 enums

DennieDan committed:
Add JavaDoc for the task types

DennieDan committed:
Add JavaDoc to Parsers classes

DennieDan committed:
Add find-with-1-keyword feature

DennieDan committed:
Remove InvalidRequestException.java

DennieDan committed:
Remove TodoWrongFormatException.java

DennieDan committed:
Add JavaDoc for CommandScanner and Storage

DennieDan committed:
Add JavaDoc for JsonLoadingException and JsonLoadingExceptionType

DennieDan committed:
Fix DateTimeParseException exception bug

DennieDan committed:
Change final attribute SCANNER to screaming-snake-case

DennieDan committed:
Merge branch 'branch-A-JavaDoc'

DennieDan committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/tecna/CommandScanner.java

DennieDan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/tecna/CommandScanner.java

DennieDan committed:
Refactor data file path for JAR access

DennieDan committed:
Merge branch 'branch-file-path'

Gra7ityIC3 commented on src/main/java/tecna/CommandScanner.java:
```diff
@@ -0,0 +1,115 @@
+ * @author DennieDan.
+ */
+public class CommandScanner {
+    private final Scanner SCANNER;
```
Should `SCANNER` be capitalized?

Gra7ityIC3 commented on src/main/java/tecna/Storage.java:
```diff
@@ -0,0 +1,133 @@
+     * https://crunchify.com/how-to-write-json-object-to-file-in-java/.
+     */
+    @SuppressWarnings("unchecked")
+    public void save(TaskList taskList) throws IOException {
```
Any reason why you chose to save the tasks in JSON instead of plain text?

DennieDan replied:
I think JSON has its format with title and value, it would be easier to parse and maintain the validity of the data file also.

Gra7ityIC3 commented on src/main/java/tecna/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private int size;
```
Perhaps you could use `tasks.size()` instead of maintaining a separate variable?

Gra7ityIC3 commented on src/main/java/tecna/CommandScanner.java:
```diff
@@ -0,0 +1,115 @@
+    public CommandType getRequest() {
+        this.input = this.SCANNER.nextLine().trim();
+        String[] input_words = this.input.split(" ");
+        if (input.equalsIgnoreCase("bye")) {
```
Perhaps you could use `input_words[0]` for the comparison here (and line 55), ignoring extraneous parameters? That way, even if the user enters `bye 123`, it will be interpreted as `bye`.

Then you could use a `switch` statement which would make your code a lot neater. :smiley:

DennieDan replied:
I see what you mean, thanks

Gra7ityIC3 commented on src/test/java/tecna/CommandScannerTest.java:
Good job on the adequate test cases! 👍

Gra7ityIC3 reviewed (commented):
For your Javadoc comments, there should be an empty line between the description and parameter section.

Also, I think the `this` keyword can be omitted when your variable isn't shadowed by a method or constructor parameter.

DennieDan committed:
Add JavaFX design

DennieDan committed:
Integrate GUI to the app

DennieDan committed:
Fix Tecna avatar

DennieDan committed:
Add closing app after "bye" command

DennieDan committed:
Merge branch 'branch-Level-10'

PrishaVP commented on src/main/java/tecna/CommandScanner.java:
```diff
@@ -0,0 +1,162 @@
+     * Gets the user's input and parses into the chatbot's valid type command.
+     * @return the corresponding command type in <code>CommandType enum</code>.
+     */
+    public CommandType getRequest() {
```
the naming feels a bit vague as it is not really clear what Request means. Especially because the input from the user is referred to as 'input' in other places in your code. Maybe can change the method name to getUserRequest()  or something like that to make it more obvious what the purpose of this method is

PrishaVP commented on src/main/java/tecna/CommandScanner.java:
```diff
@@ -0,0 +1,162 @@
+     * Generates the required index and stores in the <code>inputIndex</code> attribute.
+     * @return <code>true</code> if the parsing process is success and <code>false</code> otherwise.
+     */
+    public boolean generateIndex() {
```
i like the naming here as it is very clear what the method does

PrishaVP commented on src/main/java/tecna/DialogBox.java:
```diff
@@ -0,0 +1,70 @@
+ * Represents a dialog box consisting of an ImageView to represent the speaker's face
+ * and a label containing text from the speaker.
+ */
+public class DialogBox extends HBox {
```
maybe can fix the spelling of Dialogue (below also). because the naming standards advises full spelling rather than texting-style/casual spelling

PrishaVP commented on src/main/java/tecna/Main.java:
```diff
@@ -0,0 +1,37 @@
+    private Tecna tecna = new Tecna("data/tecna.json");
+
+    @Override
+    public void start(Stage stage) {
```
could make the name more precise by naming it startApplication or something like that but start is still quite clear to me so if you feel like putting extra effort then can be more precise i guess

PrishaVP commented on src/main/java/tecna/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    private ArrayList<Task> tasks;
```
i like that you made it plural

PrishaVP reviewed (commented):
naming LGTM except for a few minor suggestions

DennieDan committed:
Add greeting message to GUI

All printing methods in UI class is modified from void to String so that
the String can be passed to the MainWindow class and generate a new
DialogBox

DennieDan committed:
Change jar file's name

DennieDan committed:
Include assertions for parameters assumptions

DennieDan committed:
Fix readability in Tecna and Storage

DennieDan committed:
Refactor parser package

Move all JSON parser classes into the parser package

DennieDan committed:
Refactor ui package

Move Ui class into ui package.

DennieDan committed:
Refactor storage package

DennieDan committed:
Refactor task and collection packages

All the Task classes are put into "task" while TaskList is put into
"collection".

DennieDan committed:
Refactor exception package

DennieDan committed:
Refactor controller package

DennieDan committed:
Refactor command package

DennieDan committed:
Create ToDoCommand class

DennieDan committed:
Add DeadlineCommand class

DennieDan committed:
Add EventCommand class

DennieDan committed:
Clean code in DeadlineCommand

Use the parseDateTime() method in util instead of parsing repeatedly.

DennieDan committed:
Add ByeCommand class

DennieDan committed:
Add ListCommand class

DennieDan committed:
Add MarkCommand class

DennieDan committed:
Add UnmarkCommand class

DennieDan committed:
Add DeleteCommand class

DennieDan committed:
Add FindCommand class

DennieDan committed:
Add InvalidCommand class

DennieDan committed:
Add duplicate detect based on type of task

DennieDan committed:
Refactor command parsing methods

DennieDan committed:
Merge pull request #1 from DennieDan/branch-A-CodeQuality

Improve Code Quality

DennieDan committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/tecna/CommandScanner.java
#	src/main/java/tecna/Tecna.java
#	src/main/java/tecna/collection/TaskList.java
#	src/main/java/tecna/parser/EventParser.java

DennieDan committed:
Merge branch 'branch-C-DetectDuplicates'

# Conflicts:
#	src/main/java/tecna/Tecna.java
#	src/main/java/tecna/collection/TaskList.java

DennieDan committed:
Add toString() for TaskList using Java stream

DennieDan committed:
Modify findTasks method

Stream is used to improve the code.

DennieDan committed:
Merge branch 'branch-A-Streams'

DennieDan committed:
Add User Guide draft 1

DennieDan committed:
Fix repeating bug in ui

DennieDan committed:
Merge branch 'branch-fix-DetectDuplicates-bug'

DennieDan committed:
Add an "find magic" image to User Guide

DennieDan committed:
Tailor UG with images and product introduction

DennieDan committed:
Fix minor typo

DennieDan committed:
Merge branch 'branch-A-UserGuide'

# Conflicts:
#	src/main/resources/data/tecna1.json

DennieDan committed:
Change backup datafile

DennieDan committed:
Fix Things to note appearance on website

It is currently not shown correctly.

DennieDan committed:
Upload Ui.png for app website

DennieDan committed:
Rename Ui.png file

DennieDan committed:
Refactor website images

DennieDan committed:
Fix to accept "todo" without any spaces

DennieDan committed:
Fix empty task name handling for ToDo and Deadline

DennieDan committed:
Add EventCommand test cases

DennieDan committed:
Add error handling to EventCommand

DennieDan committed:
Fix MarkCommand bug

"mark   " is now handled.

DennieDan committed:
Fix UnmarkCommand bug

"unmark", "unmark " and "unmark     " are now handled

DennieDan committed:
Fix DeleteCommand bug

DennieDan committed:
Merge branch 'branch-fix-InputHandling-bug'

DennieDan committed:
Update Sources to User Guide

DennieDan committed:
Modify jar file name

DennieDan committed:
Merge branch 'branch-TestInLinux'

DennieDan committed:
Refactor data directory

Data files in the class project cannot be written to.

DennieDan committed:
Fix load and save bugs

DennieDan committed:
Modify save() method

Use a File object instead of a filePath.

DennieDan committed:
Fix DeleteCommand bug

Delete twice.

DennieDan committed:
Merge branch 'branch-SaveData'

DennieDan committed:
Fix dialog box width

DennieDan committed:
Increase text size in DialogBox

DennieDan committed:
Modify User Guide

DennieDan committed:
Merge branch 'branch-fixGUI'

DennieDan committed:
Add PauseTransition to app closing

DennieDan committed:
Fix /from can only before /to

DennieDan committed:
Merge branch 'branch-fix-EventCommand-format'

DennieDan committed:
Add JavaDoc for collection and command packages

DennieDan committed:
Add JavaDoc for controller package

DennieDan committed:
Add JavaDoc for "exception" package

DennieDan committed:
Add JavaDoc for "parser" package

DennieDan committed:
Add JavaDoc for storage package

DennieDan committed:
Add JavaDoc for "task" package

DennieDan committed:
Add JavaDoc for ui package

DennieDan committed:
Add JavaDoc for util package

DennieDan committed:
Merge branch 'branch-add-JavaDoc'

DennieDan committed:
Change Tecna's avatar to free use source

DennieDan committed:
Give credit to json-simple

DennieDan committed:
Give credit to peers

DennieDan committed:
Test regressions

DennieDan committed:
Merge branch 'branch-giving-credit'

DennieDan committed:
Fix DeadlineCommand regressions

DennieDan committed:
Fix alert errors
